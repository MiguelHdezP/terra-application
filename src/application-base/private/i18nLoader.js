/* eslint-disable global-require, import/no-named-as-default */
import { shouldPolyfill as shouldPolyfillGetCanonicalLocales } from '@formatjs/intl-getcanonicallocales/should-polyfill';
import { shouldPolyfill as shouldPolyfillPluralRules } from '@formatjs/intl-pluralrules/should-polyfill';
import { shouldPolyfill as shouldPolyfillRelativeTimeFormat } from '@formatjs/intl-relativetimeformat/should-polyfill';
import { shouldPolyfill as shouldPolyfillNumberFormat } from '@formatjs/intl-numberformat/should-polyfill';
import { shouldPolyfill as shouldPolyfillDateTimeFormat } from '@formatjs/intl-datetimeformat/should-polyfill';
import loadIntl from './intlLoaders';
import loadTranslations from './translationsLoaders';

const permitParams = (callback) => {
  if (typeof (callback) !== 'function') {
    throw new Error('Second argument must be function');
  }
};

const i18nLoader = async (locale, callback) => {
  permitParams(callback);

  const polyfills = {
    getcanonicallocales: {
      shouldPolyfill: shouldPolyfillGetCanonicalLocales,
      shouldLoadLocaleData: false, // getCononicalLocales does not have locale data
    },
    pluralrules: {
      shouldPolyfill: shouldPolyfillPluralRules,
      shouldLoadLocaleData: Intl.PluralRules.polyfilled,
    },
    relativetimeformat: {
      shouldPolyfill: shouldPolyfillRelativeTimeFormat,
      shouldLoadLocaleData: Intl.RelativeTimeFormat.polyfilled,
    },
    numberformat: {
      shouldPolyfill: shouldPolyfillNumberFormat,
      shouldLoadLocaleData: Intl.NumberFormat.polyfilled,
    },
    datetimeformat: {
      shouldPolyfill: shouldPolyfillDateTimeFormat,
      shouldLoadLocaleData: Intl.DateTimeFormat.polyfilled,
    },
  };

  Object.keys(polyfills).map(async key => {
    try {
      if (polyfills[key].shouldPolyfill()) {
        const polyfill = `@formatjs/intl-${key}/polyfill`;
        await import(polyfill);
      }
    } catch (error) {
      const polyfill = `@formatjs/intl-${key}/polyfill`;
      await import(polyfill);
    }

    if (polyfills[key].shouldLoadLocaleData) {
      loadIntl(locale, key);
    }
  });

  loadTranslations(locale, callback);
};

export default i18nLoader;
/* eslint-enable global-require */
