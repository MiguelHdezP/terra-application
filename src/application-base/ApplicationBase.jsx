/* global TERRA_THEME_CONFIG */

import React, {
  useRef, useEffect, useMemo,
} from 'react';
import PropTypes from 'prop-types';
import Base from 'terra-base';
import ThemeProvider from 'terra-theme-provider';
import { ActiveBreakpointProvider } from 'terra-breakpoints';
import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';

import { ApplicationIntlProvider } from '../application-intl';
import { NavigationPromptCheckpoint } from '../navigation-prompt';

import getBrowserLocale from './private/getBrowserLocale';
import useTestOverrides from './private/useTestOverrides';
import './private/initializeInert';

const browserLocale = getBrowserLocale();

// We only need to retrieve the root theme and root theme name once for the life of the application.
const themeConfig = (typeof (TERRA_THEME_CONFIG) !== 'undefined') ? TERRA_THEME_CONFIG : undefined;
const rootThemeName = themeConfig?.theme ? themeConfig.theme : 'terra-default-theme';

const propTypes = {
  /**
   * The components to render within ApplicationBase.
   */
  children: PropTypes.node.isRequired,
  /**
   * The locale name to be used to load translated messages.
   * If the `locale` prop is not provided, the preferred language from the browser will be used.
   */
  locale: PropTypes.string,
  /**
   * The name of the theme to apply to the application using terra-theme-provider.
   */
  themeName: PropTypes.string,
  /**
   * By default, NavigationPrompts rendered within ApplicationBase will cause the user to be prompted during
   * the window's beforeUnload event. If `unloadPromptIsDisabled` is provided, the user will **not** be prompted
   * before continuing with the unload event, even if NavigationPrompts are present.
   */
  unloadPromptIsDisabled: PropTypes.bool,
};

const ApplicationBase = ({
  locale, themeName, children, unloadPromptIsDisabled,
}) => {
  const registeredPromptsRef = useRef();

  const { localeOverride } = useTestOverrides(); // Allows us to test deployed applications in different locales.

  const theme = useMemo(() => ({
    // If the theme class name is undefined or an empty string, that indicates we have the root theme and should apply the root theme name.
    name: themeName || rootThemeName,
    className: themeName,
  }), [themeName]);

  useEffect(() => {
    if (unloadPromptIsDisabled) {
      return undefined;
    }

    function onBeforeUnload(event) {
      if (registeredPromptsRef.current && registeredPromptsRef.current.length) {
        event.preventDefault();

        // Chrome requires returnValue to be set to present the confirmation dialog
        event.returnValue = ''; // eslint-disable-line no-param-reassign

        // For this prompt, ApplicationBase is limited to browser-defaulted messaging.
        return '';
      }

      return undefined;
    }

    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [unloadPromptIsDisabled]);

  return (
    <ThemeProvider themeName={themeName}>
      <ThemeContextProvider theme={theme}>
        <Base locale={localeOverride || locale || browserLocale}>
          <ApplicationIntlProvider>
            <ActiveBreakpointProvider>
              <NavigationPromptCheckpoint
                onPromptChange={(registeredPrompts) => {
                  registeredPromptsRef.current = registeredPrompts;
                }}
              >
                {children}
              </NavigationPromptCheckpoint>
            </ActiveBreakpointProvider>
          </ApplicationIntlProvider>
        </Base>
      </ThemeContextProvider>
    </ThemeProvider>
  );
};

ApplicationBase.propTypes = propTypes;

export default ApplicationBase;
