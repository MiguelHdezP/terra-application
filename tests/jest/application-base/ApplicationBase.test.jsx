import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import ApplicationBase from '../../../src/application-base';
import i18nLoader from '../../../src/application-base/private/i18nLoader';
import logger from '../../../src/utils/logger';
import useTestOverrides from '../../../src/application-base/private/useTestOverrides';

jest.mock('../../../src/application-base/private/i18nLoader', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../../src/utils/logger', () => ({
  __esModule: true,
  default: {
    error: jest.fn(),
  },
}));

jest.mock('../../../src/application-base/private/useTestOverrides', () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe('ApplicationBase', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('renders null while no messages have loaded', () => {
    i18nLoader.mockImplementation(() => Promise.resolve({ default: undefined }));
    useTestOverrides.mockImplementation(() => ({}));

    render((
      <ApplicationBase>
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    expect(screen.queryByTestId('test-child')).toBe(null);
    expect(i18nLoader).toHaveBeenCalledWith('en');
  });

  test('renders children after messages have loaded', async () => {
    i18nLoader.mockImplementation(() => Promise.resolve({ default: { test: 'message' } }));
    useTestOverrides.mockImplementation(() => ({}));

    render((
      <ApplicationBase>
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    const child = await screen.findByTestId('test-child');

    expect(child).toBeInTheDocument();
    expect(i18nLoader).toHaveBeenCalledWith('en');
  });

  test('throws if messages could not be loaded for locale', async () => {
    const testError = new Error('Test Error');
    i18nLoader.mockImplementation(() => Promise.reject(testError));
    useTestOverrides.mockImplementation(() => ({}));

    render((
      <ApplicationBase>
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    await waitFor(() => expect(logger.error).toHaveBeenCalledWith(testError));
  });

  test('loads messages for provided locale', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    i18nLoader.mockImplementation(() => Promise.resolve({ default: { test: 'message' } }));
    useTestOverrides.mockImplementation(() => ({}));

    render((
      <ApplicationBase locale="fr">
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    const child = await screen.findByTestId('test-child');

    expect(child).toBeInTheDocument();
    expect(i18nLoader).toHaveBeenCalledWith('fr');
  });

  test('loads messages for overridden locale', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    i18nLoader.mockImplementation(() => Promise.resolve({ default: { test: 'message' } }));
    useTestOverrides.mockImplementation(() => ({ localeOverride: 'es' }));

    render((
      <ApplicationBase locale="fr">
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    const child = await screen.findByTestId('test-child');

    expect(child).toBeInTheDocument();
    expect(i18nLoader).toHaveBeenCalledWith('es');
  });

  test('adds appropriate theme to document', async () => {
    i18nLoader.mockImplementation(() => Promise.resolve({ default: { test: 'message' } }));
    useTestOverrides.mockImplementation(() => ({ localeOverride: 'es' }));

    const view = render((
      <ApplicationBase themeName="test-theme">
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    await waitFor(() => expect(document.documentElement.classList.contains('test-theme')).toBeTruthy());

    view.rerender((
      <ApplicationBase themeName="test-theme-2">
        <div data-testid="test-child" />
      </ApplicationBase>
    ));

    await waitFor(() => expect(document.documentElement.classList.contains('test-theme')).toBeFalsy());
    await waitFor(() => expect(document.documentElement.classList.contains('test-theme-2')).toBeTruthy());
  });
});

