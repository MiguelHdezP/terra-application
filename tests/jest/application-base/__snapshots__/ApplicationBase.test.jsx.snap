// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationBase should render with all props 1`] = `
<ThemeProvider
  themeName="test-theme"
>
  <ThemeContextProvider
    theme={
      Object {
        "className": "test-theme",
        "name": "test-theme",
      }
    }
  >
    <Base
      customMessages={Object {}}
      locale="en"
      strictMode={false}
    >
      <InjectIntl(Component)>
        <ActiveBreakpointProvider>
          <withPromptRegistration(NavigationPromptCheckpoint)
            onPromptChange={[Function]}
          >
            <div>
              content
            </div>
          </withPromptRegistration(NavigationPromptCheckpoint)>
        </ActiveBreakpointProvider>
      </InjectIntl(Component)>
    </Base>
  </ThemeContextProvider>
</ThemeProvider>
`;

exports[`ApplicationBase should render with minimal props 1`] = `
<ThemeProvider>
  <ThemeContextProvider
    theme={
      Object {
        "className": undefined,
        "name": "terra-default-theme",
      }
    }
  >
    <Base
      customMessages={Object {}}
      locale="en"
      strictMode={false}
    >
      <InjectIntl(Component)>
        <ActiveBreakpointProvider>
          <withPromptRegistration(NavigationPromptCheckpoint)
            onPromptChange={[Function]}
          >
            <div>
              content
            </div>
          </withPromptRegistration(NavigationPromptCheckpoint)>
        </ActiveBreakpointProvider>
      </InjectIntl(Component)>
    </Base>
  </ThemeContextProvider>
</ThemeProvider>
`;

exports[`ApplicationBase should render with the preferred browser local 1`] = `
<ThemeProvider>
  <ThemeContextProvider
    theme={
      Object {
        "className": undefined,
        "name": "terra-default-theme",
      }
    }
  >
    <Base
      customMessages={Object {}}
      locale="en"
      strictMode={false}
    >
      <InjectIntl(Component)>
        <ActiveBreakpointProvider>
          <withPromptRegistration(NavigationPromptCheckpoint)
            onPromptChange={[Function]}
          >
            <div>
              content
            </div>
          </withPromptRegistration(NavigationPromptCheckpoint)>
        </ActiveBreakpointProvider>
      </InjectIntl(Component)>
    </Base>
  </ThemeContextProvider>
</ThemeProvider>
`;
